django/db/migrations/migration.py

```diff
@@ -108,20 +117,41 @@ def apply(self, project_state, schema_editor, collect_sql=False):
                 schema_editor.collected_sql.append("-- %s" % operation.describe())
                 schema_editor.collected_sql.append("--")
                 if not operation.reduces_to_sql:
-                    continue
-            # Save the state before the operation has run
-            old_state = project_state.clone()
-            operation.state_forwards(self.app_label, project_state)
+                    return (from_state, to_state)
+            operation.state_forwards(self.app_label, to_state)
             # Run the operation
             atomic_operation = operation.atomic or (self.atomic and operation.atomic is not False)
             if not schema_editor.atomic_migration and atomic_operation:
                 # Force a transaction on a non-transactional-DDL backend or an
                 # atomic operation inside a non-atomic migration.
                 with atomic(schema_editor.connection.alias):
-                    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
+                    operation.database_forwards(self.app_label, schema_editor, from_state, to_state)
             else:
                 # Normal behaviour
-                operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
+                operation.database_forwards(self.app_label, schema_editor, from_state, to_state)
+            return (from_state, to_state)
+
+        def _apply_operations(operations):
+            if not operations:
+                return
+
+            for operation in operations:
```

Comment:
@charettes We could pretty easily add a `emit_pre_operation_signal` here (and in `unapply`). I think it's better to keep it out of this PR since I don't know about any use case for it but thought it would be worth mentioning.

Reply:
@arthurio yeah I'd suggest we stick to only `post_operation` for now because we have a use case for it. Adding a _pre_ signal could be useful but adding features for the sake of it only complexify the implementation for unknown benefits.